File: app/src/main/java/com/krystianwsul/checkme/firebase/records/RemoteRecord.java
package com.krystianwsul.checkme.firebase.records;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.Log;

import junit.framework.Assert;

import java.util.HashMap;
import java.util.Map;

abstract public class RemoteRecord {
    boolean mDelete = false;
    boolean mDeleted = false;

    final boolean mCreate;
    boolean mCreated = false;

    final Map<String, Object> mUpdate;
    boolean mUpdated = false;

    RemoteRecord(boolean create) {
        if (create) {
            mCreate = true;
            mUpdate = null;
        } else {
            mCreate = false;
            mUpdate = new HashMap<>();
        }
    }

    @NonNull
    protected abstract String getKey();

    @NonNull
    protected abstract Object getCreateObject();

    void getValues(@NonNull Map<String, Object> values) {
        Assert.assertTrue(!mDeleted);
        Assert.assertTrue(!mCreated);
        Assert.assertTrue(!mUpdated);

        if (mDelete) {
            Log.e("asdf", "RemoteRecord.getValues deleting " + this);

            Assert.assertTrue(!mCreate);
            Assert.assertTrue(mUpdate != null);

            mDeleted = true;
            values.put(getKey(), null);
        } else if (mCreate) {
            Log.e("asdf", "RemoteRecord.getValues creating " + this);

            Assert.assertTrue(mUpdate == null);

            mCreated = true;
            values.put(getKey(), getCreateObject());
        } else {
            Assert.assertTrue(mUpdate != null);

            if (!mUpdate.isEmpty()) {
                Log.e("asdf", "RemoteRecord.getValues updating " + this);

                mUpdated = true;
                values.putAll(mUpdate);
            }
        }
    }

    void addValue(@NonNull String key, @Nullable Object object) {
        Assert.assertTrue(!mDelete);
        Assert.assertTrue(!mDeleted);
        Assert.assertTrue(!mCreated);
        Assert.assertTrue(!mUpdated);

        if (mCreate) {
            Assert.assertTrue(mUpdate == null);
        } else {
            Assert.assertTrue(mUpdate != null);

            mUpdate.put(key, object);
        }
    }

    public void delete() {
        Assert.assertTrue(!mDeleted);
        Assert.assertTrue(!mUpdated);
        Assert.assertTrue(!mCreated);
        Assert.assertTrue(!mDelete);
        Assert.assertTrue(!mCreate);
        Assert.assertTrue(mUpdate != null);

        mDelete = true;
    }
}

File: app/src/main/java/com/krystianwsul/checkme/firebase/records/RemoteRecord.kt
package com.krystianwsul.checkme.firebase.records

import android.util.Log


abstract class RemoteRecord(create: Boolean) {

    protected var delete = false

    var update = if (create) null else mutableMapOf<String, Any?>()

    abstract val key: String

    abstract val createObject: Any

    open val children = listOf<RemoteRecord>()

    protected abstract fun deleteFromParent()

    fun getValues(values: MutableMap<String, Any?>): Boolean {
        if (delete) {
            Log.e("asdf", "RemoteRecord.getValues deleting $this")

            check(update != null)

            values[key] = null
            delete = false

            deleteFromParent()

            return true
        } else {
            if (update == null) {
                Log.e("asdf", "RemoteRecord.getValues creating $this")

                values[key] = createObject

                setCreated()
            } else {
                if (update!!.isNotEmpty()) {
                    Log.e("asdf", "RemoteRecord.getValues updating $this")

                    values.putAll(update!!)

                    update = mutableMapOf()
                }

                children.forEach { it.getValues(values) }
            }

            return false
        }
    }

    private fun setCreated() {
        check(update == null)

        update = mutableMapOf()

        children.forEach { it.setCreated() }
    }

    protected fun addValue(key: String, obj: Any?) {
        check(!delete)

        update?.put(key, obj)
    }

    fun delete() {
        check(!delete)
        check(update != null)

        delete = true
    }
}
